runtime/lua/vim/treesitter/query.lua

```diff
@@ -114,9 +128,28 @@ local explicit_queries = setmetatable({}, {
 ---
 --- @param lang string: The language to use for the query
 --- @param query_name string: The name of the query (i.e. "highlights")
---- @param text string: The query text (unparsed).
+--- @param text string|function: The query text (unparsed) or function returning the text.
 function M.set_query(lang, query_name, text)
-  explicit_queries[lang][query_name] = M.parse_query(lang, text)
+  explicit_queries[lang][query_name] = text
```

Comment:
get_query parses the queries every time anyways when they come from the runtime files. get_query needs to be cached anyways as reloading the queries was often too slow for most users. But we could also keep it